File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/command/memento_solution/Driver.java
// Driver.java

public class Driver
{
	public Driver(String aName)
	{
		setName(aName);
		setX(0);
		setY(0);
		setHeading(NORTH);
	}

	public Driver(Driver aDriver)
	{
		copy(aDriver);
	}

	private void copy(Driver aDriver)
	{
		setName(aDriver.getName());
		setX(aDriver.getX());
		setY(aDriver.getY());
		setHeading(aDriver.getHeading());
	}

	public void turnLeft()
	{
		switch(getHeading())
		{
			case NORTH: setHeading(WEST);  break;
			case EAST:  setHeading(NORTH); break;
			case SOUTH: setHeading(EAST);  break;
			case WEST:  setHeading(SOUTH); break;
		}
	}

	public void goStraight(int howFar)
	{
		switch(getHeading())
		{
			case NORTH: setY(getY() + howFar);  break;
			case EAST:  setX(getX() + howFar); break;
			case SOUTH: setY(getY() - howFar);  break;
			case WEST:  setX(getX() - howFar); break;
		}
	}

	public void turnRight()
	{
		switch(getHeading())
		{
			case NORTH: setHeading(EAST);  break;
			case EAST:  setHeading(SOUTH); break;
			case SOUTH: setHeading(WEST);  break;
			case WEST:  setHeading(NORTH); break;
		}
	}

	public void undoTurnLeft()
	{
		switch(getHeading())
		{
			case WEST:  setHeading(NORTH); break;
			case NORTH: setHeading(EAST);  break;
			case EAST:  setHeading(SOUTH); break;
			case SOUTH: setHeading(WEST);  break;
		}
	}

	public void undoGoStraight(int howFar)
	{
		switch(getHeading())
		{
			case NORTH: setY(getY() - howFar);  break;
			case EAST:  setX(getX() - howFar); break;
			case SOUTH: setY(getY() + howFar);  break;
			case WEST:  setX(getX() + howFar); break;
		}
	}

	public void undoTurnRight()
	{
		switch(getHeading())
		{
			case EAST:  setHeading(NORTH); break;
			case SOUTH: setHeading(EAST);  break;
			case WEST:  setHeading(SOUTH); break;
			case NORTH: setHeading(WEST);  break;
		}
	}

	public void reportPosition()
	{
		System.out.println(
			getName()
		+	" is located at ("
		+	getX()
		+	", "
		+	getY()
		+	") facing "
		+	getHeadingAsString()
		+	".");
	}

	public Driver clone()
	{
		return new Driver(this);
	}

	private void setName(String aName)
	{
		myName = aName;
	}

	private String getName()
	{
		return myName;
	}

	private void setX(int aX)
	{
		myX = aX;
	}

	private int getX()
	{
		return myX;
	}

	private void setY(int aY)
	{
		myY = aY;
	}

	private int getY()
	{
		return myY;
	}

	private void setHeading(int aHeading)
	{
		myHeading = aHeading;
	}

	private int getHeading()
	{
		return myHeading;
	}

	private String getHeadingAsString()
	{
		String result = null;

		switch(getHeading())
		{
			case NORTH: result = "NORTH"; break;
			case EAST:  result = "EAST" ; break;
			case SOUTH: result = "SOUTH"; break;
			case WEST:  result = "WEST" ; break;
		}

		return result;
	}

	private static final int NORTH = 1;
	private static final int EAST  = 2;
	private static final int SOUTH = 3;
	private static final int WEST  = 4;

	private String myName;
	private int myX;
	private int myY;
	private int myHeading;

	public MementoInterface backup()
	{
		return new MementoImplementation(this);
	}

	public void restore(MementoInterface aMementoInterface)
	{
		if (aMementoInterface == null)
		{
			throw new DpException("void Driver.restore(MementoInterface aMementoInterface) -- aMementoInterface == null.");
		}

		if (!(aMementoInterface instanceof MementoImplementation))
		{
			throw new DpException("void Driver.restore(MementoInterface aMementoInterface) -- that's the wrong kind of memento.");
		}

		MementoImplementation aMementoImplementation =
			(MementoImplementation) aMementoInterface;

		if (aMementoImplementation.myOriginator != this)
		{
			throw new DpException("void Driver.restore(MementoInterface aMementoInterface) -- that memento didn't come from me.");
		}

		copy(aMementoImplementation.myClone);
	}

	public static interface MementoInterface
	{
	}

	private static class MementoImplementation implements MementoInterface
	{
		public MementoImplementation(Driver aOriginator)
		{
			myOriginator = aOriginator;
			myClone = aOriginator.clone();
		}

		public Driver myOriginator;
		public Driver myClone;
	}
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/command/lab/driver.cpp
// driver.cpp

#include <iostream>
#include <string>
#include "dp_exception.h"
#include "driver.h"

driver::driver(const std::string & a_name)
{
	set_name(a_name);
	set_x(0);
	set_y(0);
	set_heading(NORTH);
}

void driver::turn_left()
{
	switch(get_heading())
	{
	case NORTH: set_heading(WEST);  break;
	case EAST:  set_heading(NORTH); break;
	case SOUTH: set_heading(EAST);  break;
	case WEST:  set_heading(SOUTH); break;
	}
}

void driver::go_straight(int how_far)
{
	switch(get_heading())
	{
	case NORTH: set_y(get_y() + how_far);  break;
	case EAST:  set_x(get_x() + how_far); break;
	case SOUTH: set_y(get_y() - how_far);  break;
	case WEST:  set_x(get_x() - how_far); break;
	}
}

void driver::turn_right()
{
	switch(get_heading())
	{
	case NORTH: set_heading(EAST);  break;
	case EAST:  set_heading(SOUTH); break;
	case SOUTH: set_heading(WEST);  break;
	case WEST:  set_heading(NORTH); break;
	}
}

void driver::undo_turn_left()
{
	switch(get_heading())
	{
	case NORTH: set_heading(EAST);  break;
	case EAST:  set_heading(SOUTH); break;
	case SOUTH: set_heading(WEST);  break;
	case WEST:  set_heading(NORTH); break;
	}
}

void driver::undo_go_straight(int how_far)
{
	switch(get_heading())
	{
	case NORTH: set_y(get_y() - how_far);  break;
	case EAST:  set_x(get_x() - how_far); break;
	case SOUTH: set_y(get_y() + how_far);  break;
	case WEST:  set_x(get_x() + how_far); break;
	}
}

void driver::undo_turn_right()
{
	switch(get_heading())
	{
	case NORTH: set_heading(WEST);  break;
	case EAST:  set_heading(NORTH); break;
	case SOUTH: set_heading(EAST);  break;
	case WEST:  set_heading(SOUTH); break;
	}
}

void driver::report_position() const
{
	std::cout
		<< get_name()
		<< " is located at ("
		<< get_x()
		<< ", "
		<< get_y()
		<< ") facing "
		<< get_heading_as_string()
		<< "."
		<< std::endl;
}

driver * driver::clone() const
{
	return new driver(*this);
}

void driver::set_name(const std::string & a_name)
{
	my_name = a_name;
}

const std::string & driver::get_name() const
{
	return my_name;
}

void driver::set_x(int a_x)
{
	my_x = a_x;
}

int driver::get_x() const
{
	return my_x;
}

void driver::set_y(int a_y)
{
	my_y = a_y;
}

int driver::get_y() const
{
	return my_y;
}

void driver::set_heading(int a_heading)
{
	my_heading = a_heading;
}

int driver::get_heading() const
{
	return my_heading;
}

std::string driver::get_heading_as_string() const
{
	std::string result;

	switch(get_heading())
	{
	case NORTH: result = "NORTH"; break;
	case EAST:  result = "EAST" ; break;
	case SOUTH: result = "SOUTH"; break;
	case WEST:  result = "WEST" ; break;
	}

	return result;
}

driver::memento_interface * driver::backup()
{
	return new memento_implementation(*this);
}

void driver::restore(memento_interface * a_memento_interface)
{
	memento_implementation * a_memento_implementation =
		dynamic_cast<memento_implementation *>(a_memento_interface);

	if (a_memento_implementation == nullptr)
	{
		throw dp_exception("void driver::restore(memento_interface * a_memento_interface) -- that's the wrong kind of memento.");
	}

	if (a_memento_implementation->my_originator != this)
	{
		throw dp_exception("void driver::restore(memento_interface * a_memento_interface) -- that memento didn't come from me.");
	}

	*this = *a_memento_implementation->my_clone;
}

driver::memento_interface::memento_interface()
{
}

driver::memento_interface::~memento_interface()
{
}

driver::memento_implementation::memento_implementation(driver & a_originator)
{
	my_originator = &a_originator;
	my_clone      = a_originator.clone();
}
